File: bindings/lua/src/ccharts.lua
-- ccharts.lua — high-level Lua facade for the ccharts charting library.
--
-- This is a thin, pure-Lua layer over the compiled C module (ccharts_core,
-- built against the vendored C ABI). It owns the fiddly color-name resolution
-- (color names -> ANSI escapes) and the friendly, table-based API; the C
-- module owns all struct/array marshalling (compiling the vendored ABI and
-- building contiguous struct blocks for pie/bar/stack/box).

local core = require "ccharts_core"

local ccharts = {}

-- Color name -> ccharts_color() index (mirrors ccharts_color_index in the ABI).
local COLOR = {
  black = 0, red = 1, green = 2, yellow = 3, blue = 4, magenta = 5,
  cyan = 6, white = 7, bright_black = 8, bright_red = 9, bright_green = 10,
  bright_yellow = 11, bright_blue = 12, bright_magenta = 13, bright_cyan = 14,
  bright_white = 15, reset = 16,
}

-- Resolve a colour option to the ANSI escape string the C layer expects.
-- nil -> nil (library default); a colour name -> its escape; anything else
-- (already an escape string) is passed through.
local function resolve(v)
  if v == nil then return nil end
  if type(v) == "string" then
    local idx = COLOR[v]
    if idx ~= nil then return core._color(idx) end
    return v -- an arbitrary ANSI/truecolor escape
  end
  return tostring(v)
end

local EMPTY = "" -- plain mode: emit no escape at all (distinct from NULL)

-- ------------------------------------------------------------------
-- Dataset wrapper (line / candle / size)
-- ------------------------------------------------------------------
local data_mt = {}
data_mt.__index = data_mt

function data_mt.size(self) return core._data_len(self._h) end
data_mt.len = data_mt.size

local function chart_settings(s)
  s = s or {}
  if s.plain then
    return {
      rise_color = EMPTY, fall_color = EMPTY, bg_color = EMPTY,
      area_color = EMPTY, single_color = s.single_color or false,
      show_prices = s.show_prices or false, show_times = s.show_times or false,
    }
  end
  return {
    rise_color = resolve(s.rise_color), fall_color = resolve(s.fall_color),
    bg_color = resolve(s.bg_color), area_color = resolve(s.area_color),
    single_color = s.single_color or false, show_prices = s.show_prices or false,
    show_times = s.show_times or false,
  }
end

function data_mt.line(self, width, height, settings)
  return core._line(self._h, width, height, chart_settings(settings))
end
function data_mt.candle(self, width, height, settings)
  return core._candle(self._h, width, height, chart_settings(settings))
end

-- ------------------------------------------------------------------
-- Building a dataset
-- ------------------------------------------------------------------

-- Build a dataset from four equal-length price columns and optional epoch
-- seconds. `ts` may be nil (all timestamps unknown).
function ccharts.from_arrays(open, high, low, close, ts)
  if not open or #open == 0 then error("need at least one candle") end
  return setmetatable({ _h = core._from_arrays(open, high, low, close, ts) },
                      data_mt)
end

-- Build a dataset from the fixed-schema JSON accepted by the C layer.
function ccharts.from_json(text)
  return setmetatable({ _h = core._from_json(text) }, data_mt)
end

-- Build a dataset from CSV rows of `open,high,low,close[,timestamp]`.
function ccharts.from_csv(text, value_separator, line_separator)
  local vs = value_separator and string.byte(value_separator, 1) or 44  -- ","
  local ls = line_separator and string.byte(line_separator, 1) or 10    -- "\n"
  return setmetatable({ _h = core._from_csv(text, vs, ls) }, data_mt)
end

-- ------------------------------------------------------------------
-- Standalone renderers (no OHLC dataset)
-- ------------------------------------------------------------------

-- pie: slices = {{label=, value=}, ...}
function ccharts.pie(slices, width, height, settings)
  if not slices or #slices == 0 then error("need at least one slice") end
  local s = settings or {}
  local out = {
    donut = s.donut or false,
    show_legend = (s.show_legend == nil) and true or s.show_legend,
    show_pct = s.show_pct or false,
  }
  if s.slice_gap ~= nil then out.slice_gap = s.slice_gap end
  if s.inner_radius_ratio ~= nil then out.inner_radius_ratio = s.inner_radius_ratio end
  if s.legend_format ~= nil then out.legend_format = s.legend_format end
  if s.start_angle ~= nil then out.start_angle = s.start_angle end
  if s.counter_clockwise then out.counter_clockwise = true end
  if s.center_text then out.center_text = s.center_text end
  if s.colors then
    local resolved = {}
    for i, c in ipairs(s.colors) do resolved[i] = resolve(c) end
    out.colors = resolved
  end
  return core._pie(slices, width, height, out)
end

-- histogram: samples = list of numbers
function ccharts.histogram(samples, width, height, settings)
  if not samples or #samples == 0 then error("need at least one sample") end
  local s = settings or {}
  local rc, bg
  if s.plain then rc, bg = EMPTY, EMPTY else rc, bg = resolve(s.rise_color), resolve(s.bg_color) end
  return core._hist(samples, width, height, {
    rise_color = rc, bg_color = bg, bin_count = s.bin_count or 0,
    min_value = (s.min_value ~= nil and s.min_value) or (0 / 0), -- NaN = auto
    max_value = (s.max_value ~= nil and s.max_value) or (0 / 0),
    show_bins = s.show_bins or false, show_prices = s.show_prices or false,
  })
end

-- sparkline: samples = list of numbers
function ccharts.sparkline(samples, width, height, settings)
  if not samples or #samples == 0 then error("need at least one sample") end
  local s = settings or {}
  local rc, ac
  if s.plain then rc, ac = EMPTY, EMPTY else rc, ac = resolve(s.rise_color), resolve(s.area_color) end
  return core._spark(samples, width, height, {
    rise_color = rc, area_color = ac,
    min_above = s.min_above or 0, min_below = s.min_below or 0,
  })
end

-- bar: items = {{label=, value=}, ...}
function ccharts.bar(items, width, height, settings)
  if not items or #items == 0 then error("need at least one bar") end
  local s = settings or {}
  local rc, bg
  if s.plain then rc, bg = EMPTY, EMPTY else rc, bg = resolve(s.rise_color), resolve(s.bg_color) end
  return core._bar(items, width, height, {
    rise_color = rc, bg_color = bg,
    show_labels = s.show_labels or false, show_prices = s.show_prices or false,
  })
end

-- stacked_bar: series = {{name=, values={...}}, ...}
function ccharts.stacked_bar(series, width, height, settings)
  if not series or #series == 0 then error("need at least one series") end
  local s = settings or {}
  local colors
  if s.plain then
    colors = {}
    for i = 1, #series do colors[i] = EMPTY end
  elseif s.colors then
    colors = {}
    for i, c in ipairs(s.colors) do colors[i] = resolve(c) end
  end
  return core._stack(series, width, height, {
    colors = colors, bg_color = (s.plain and EMPTY) or resolve(s.bg_color),
    cat_labels = s.cat_labels, show_labels = s.show_labels or false,
    show_prices = s.show_prices or false,
  })
end

-- heatmap: values = rows x cols matrix (array of equal-length arrays)
function ccharts.heatmap(values, width, height, settings)
  if not values or #values == 0 then error("need a non-empty value matrix") end
  local rows, cols = #values, #values[1]
  if cols == 0 then error("matrix columns must not be empty") end
  for r = 1, rows do
    if #values[r] ~= cols then error("all rows must have the same number of values") end
  end
  local s = settings or {}
  local function pick(k)
    if s.plain then return EMPTY end
    return resolve(s[k])
  end
  return core._heat(values, rows, cols, width, height, {
    low_color = pick("low_color"), high_color = pick("high_color"),
    mid_color = pick("mid_color"), bg_color = pick("bg_color"),
    row_labels = s.row_labels, col_labels = s.col_labels,
    show_labels = s.show_labels or false,
  })
end

-- boxplot: categories = {{name=, samples={...}}, ...}
function ccharts.boxplot(categories, width, height, settings)
  if not categories or #categories == 0 then error("need at least one category") end
  local s = settings or {}
  local rc, ac, bg
  if s.plain then rc, ac, bg = EMPTY, EMPTY, EMPTY
  else rc, ac, bg = resolve(s.rise_color), resolve(s.area_color), resolve(s.bg_color) end
  return core._box(categories, width, height, {
    rise_color = rc, area_color = ac, bg_color = bg,
    show_prices = s.show_prices or false,
  })
end

-- ------------------------------------------------------------------
-- Introspection
-- ------------------------------------------------------------------
ccharts.version = core._version()
ccharts.max_dim = core._max_dim()
ccharts.max_cells = core._max_cells()
ccharts._core = core

return ccharts

File: bindings/julia/src/Ccharts.jl
### src/Ccharts.jl — Julia binding for the ccharts single-header C library.
#
# Links the shared library produced by deps/build.jl (gcc over the vendored
# abi/ccharts_abi.c) through a ccall FFI. Struct packing mirrors the C layout
# field-for-field; contiguous struct arrays (pie/bar/stack/box) are Julia
# `Array{<:isbits struct}` blocks, which are laid out as one contiguous, C-ABI
# aligned run of `size`-byte elements — never an array of pointers.
module Ccharts

using Libdl

# Locate the built native library (deps/deps.jl), with CCHARTS_NATIVE_DIR
# fallback. Build failure here points back at `Pkg.build("Ccharts")`.
include("../deps/deps.jl")
const _LIB = Ccharts_deps.CCHARTS_LIBPATH

# ---------------------------------------------------------------------------
# Error type
# ---------------------------------------------------------------------------

"""Error raised on any non-CCHARTS_OK status from the C layer."""
struct CchartsError <: Exception
    status::Int32
    message::String
end
Base.showerror(io::IO, e::CchartsError) =
    print(io, "CchartsError(", e.status, "): ", e.message)

function _raise_on(status::Integer)
    status == 0 && return status
    msgptr = ccall((:ccharts_error_message, _LIB), Ptr{UInt8}, (Cint,), Cint(status))
    msg = msgptr == C_NULL ? "" : unsafe_string(msgptr)
    throw(CchartsError(status, msg))
end

# ---------------------------------------------------------------------------
# C settings structs (layouts from abi/ccharts_abi.h). Julia lays out isbits
# structs in C-ABI order with standard alignment/padding, so these match the
# C side byte-for-byte. All pointers are raw addresses we keep rooted via a
# `keep` vector across the render call.
# ---------------------------------------------------------------------------

# ccharts_settings (line / candle)
struct CSettings
    rise_color::Ptr{UInt8}    # UInt8 == Char in C, pointer
    fall_color::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    area_color::Ptr{UInt8}
    single_color::Int32
    show_prices::Int32
    show_times::Int32
end

# ccharts_pie_slice
struct KPieSlice
    label::Ptr{UInt8}
    value::Float64
end

# ccharts_hist_settings
struct CHistSettings
    rise_color::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    bin_count::Int32
    min_value::Float64
    max_value::Float64
    show_bins::Int32
    show_prices::Int32
end

# ccharts_spark_settings
struct CSparkSettings
    rise_color::Ptr{UInt8}
    area_color::Ptr{UInt8}
    min_above::Int32
    min_below::Int32
end

# ccharts_bar_slice
struct KBarSlice
    label::Ptr{UInt8}
    value::Float64
end

# ccharts_bar_settings
struct CBarSettings
    rise_color::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    show_labels::Int32
    show_prices::Int32
end

# ccharts_stack_series
struct KStackSeries
    name::Ptr{UInt8}
    values::Ptr{Float64}
end

# ccharts_stack_settings
struct CStackSettings
    colors::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    cat_labels::Ptr{UInt8}
    series::Int32
    cats::Int32
    show_labels::Int32
    show_prices::Int32
end

# ccharts_heat_settings
struct CHeatSettings
    low_color::Ptr{UInt8}
    high_color::Ptr{UInt8}
    mid_color::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    row_labels::Ptr{UInt8}
    col_labels::Ptr{UInt8}
    show_labels::Int32
end

# ccharts_box_category
struct KBoxCategory
    name::Ptr{UInt8}
    samples::Ptr{Float64}
    n::Int32
end

# ccharts_box_settings
struct CBoxSettings
    rise_color::Ptr{UInt8}
    area_color::Ptr{UInt8}
    bg_color::Ptr{UInt8}
    show_prices::Int32
end

# ---------------------------------------------------------------------------
# Helpers: rooted C strings + color escapes + pointer arrays.
# ---------------------------------------------------------------------------

const NUL = C_NULL  # Ptr{UInt8}(0)

# Copy a Julia string into a NUL-terminated byte buffer and root it in `keep`,
# returning the address. `nothing` -> NULL (library default).
function _cstr(s, keep)
    b = Vector{UInt8}(codeunits(s))
    push!(b, 0x00)
    p = pointer(b)
    push!(keep, b)
    return p
end

# Resolve a color setting: `nothing` -> library default (NULL); a name/escape
# -> its ANSI escape bytes; `plain=true` forces an *empty* escape so no ANSI
# is emitted at all (overriding even an explicit color).
const _ESCAPES = Dict{String,String}(
    "black"         => "\e[30m",
    "red"           => "\e[31m",
    "green"         => "\e[32m",
    "yellow"        => "\e[33m",
    "blue"          => "\e[34m",
    "magenta"       => "\e[35m",
    "cyan"          => "\e[36m",
    "white"         => "\e[37m",
    "bright_black"  => "\e[90m",
    "bright_red"    => "\e[91m",
    "bright_green"  => "\e[92m",
    "bright_yellow" => "\e[93m",
    "bright_blue"   => "\e[94m",
    "bright_magenta"=> "\e[95m",
    "bright_cyan"   => "\e[96m",
    "bright_white"  => "\e[97m",
    "reset"         => "\e[0m",
)

function _color(value, keep; plain=false)
    plain && return _cstr("", keep)         # no ANSI at all
    value === nothing && return NUL
    s = string(value)
    isempty(s) && return NUL                # empty -> library default
    _cstr(get(_ESCAPES, s, s), keep)        # name, or pass raw escape through
end

# Build a contiguous array of NUL-terminated escapes/labels (used for
# pie colors, stack colors/labels, heat labels). `terminate` appends a 0
# sentinel (stack colors). `plain` resolves like _color.
function _ptr_array(values, keep; terminate=false, plain=false)
    ptrs = Ptr{UInt8}[]
    if values === nothing
        return (NUL, ptrs)
    end
    for v in values
        push!(ptrs, _color(v, keep; plain))
    end
    terminate && push!(ptrs, NUL)
    block = Vector{Ptr{UInt8}}(ptrs)
    push!(keep, block)
    return (pointer(block), ptrs)
end

# ---------------------------------------------------------------------------
# Introspection (public)
# ---------------------------------------------------------------------------

"""Library version ("3.0.0")."""
function version()
    p = ccall((:ccharts_version, _LIB), Ptr{UInt8}, ())
    p == C_NULL ? "" : unsafe_string(p)
end

"""ANSI escape for a ccharts color index, or `nothing` when out of range."""
function color(index::Integer)
    p = ccall((:ccharts_color, _LIB), Ptr{UInt8}, (Cint,), Cint(index))
    p == C_NULL ? nothing : unsafe_string(p)
end

max_dim()    = ccall((:ccharts_max_dim, _LIB), Cint, ())
max_cells()  = ccall((:ccharts_max_cells, _LIB), Cint, ())

include("Chart.jl")

# Re-export the public surface defined in the Chart submodule. `line`/`candle`
# also work as instance methods (chart.line(w, h); ...), and the standalone
# renderers + dataset constructors are module-level functions.
using .__CchartsChart__: Chart,
    from_arrays, from_json, from_csv,
    line, candle,
    pie, histogram, sparkline, bar, stacked_bar, heatmap, boxplot

export Chart,
       from_arrays, from_json, from_csv,
       line, candle,
       pie, histogram, sparkline, bar, stacked_bar, heatmap, boxplot

export CchartsError, version, color, max_dim, max_cells

end # module